subsys/mgmt/mcumgr/smp/src/client.c

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** SMP - Simple Management Client Protocol. */
+
+#include <zephyr/kernel.h>
+#include <zephyr/init.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/net/buf.h>
+#include <zephyr/mgmt/mcumgr/mgmt/mgmt.h>
+#include <zephyr/mgmt/mcumgr/smp/smp.h>
+#include <zephyr/mgmt/mcumgr/smp/smp_client.h>
+#include <zephyr/mgmt/mcumgr/transport/smp.h>
+#include <mgmt/mcumgr/transport/smp_internal.h>
+#include <assert.h>
+#include <string.h>
+
+#include <zcbor_common.h>
+#include <zcbor_decode.h>
+#include <zcbor_encode.h>
+
+#include <zephyr/logging/log.h>
+
+#define LOG_MODULE_NAME smp_client
+LOG_MODULE_REGISTER(smp_client);
+
+#define SMP_CMD_MAX 4
+
+struct smp_client_cmd_req {
+	sys_snode_t node;
+	struct net_buf *nb;
+	void *user_data;
+	smp_client_res_fn cb;
+	int64_t timestamp;
+	int retry_cnt;
+};
+
+struct smp_client_data_base {
+	struct k_work_delayable work_delay;
+	sys_slist_t cmd_free_list;
+	sys_slist_t cmd_list;
+	/* SMP SEQ */
+	uint8_t smp_seq;
+};
+
+static struct smp_client_cmd_req smp_cmd_req_bufs[SMP_CMD_MAX];
+static struct smp_client_data_base smp_client_data;
+
+static void smp_read_hdr(const struct net_buf *nb, struct smp_hdr *dst_hdr);
+static void smp_client_cmd_req_free(struct smp_client_cmd_req *cmd_req);
+
+static void smp_header_init(struct smp_hdr *header, uint16_t group, uint8_t id, uint8_t op,
+			    uint16_t payload_len)
+{
+	/* Pre config SMP header structure */
+	memset(header, 0, sizeof(struct smp_hdr));
+	header->nh_op = op;
+	header->nh_len = sys_cpu_to_be16(payload_len);
+	header->nh_group = sys_cpu_to_be16(group);
+	header->nh_id = id;
+	header->nh_seq = smp_client_data.smp_seq++;
+}
+
+static void snmp_client_trasnport_work_fn(struct k_work *work)
+{
+	struct smp_transport *smpt;
+	struct smp_client_cmd_req *entry, *tmp;
+	smp_client_res_fn cb;
+	void *user_data;
+
+	ARG_UNUSED(work);
+
+	if (sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		/* No more packet for Transport */
+		return;
+	}
+
+	smpt = smp_transport_get();
+
+	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&smp_client_data.cmd_list, entry, tmp, node) {
+		if (smpt && k_uptime_get() - entry->timestamp < 1000) {
+			continue;
+		} else if (smpt && entry->retry_cnt < 15) {
+			/* Increment reference for re-transmission */
+			entry->nb = net_buf_ref(entry->nb);
+			entry->retry_cnt++;
+			entry->timestamp = k_uptime_get();
+			smp_tx_req(smpt, entry->nb);
+			continue;
+		}
+
+		cb = entry->cb;
+		user_data = entry->user_data;
+		smp_client_cmd_req_free(entry);
+		if (cb) {
+			cb(NULL, user_data);
+		}
+	}
+
+	if (!sys_slist_is_empty(&smp_client_data.cmd_list)) {
+		/* Re-schedule new timeout check*/
+		k_work_reschedule(&smp_client_data.work_delay, K_MSEC(50));
```

Comment:
If this does a resend every 1 second, why is it being ran every 50ms? Maybe store the next timeout and schedule it for that amount of time.

Reply:
Good Point. Let me re-think this.